lib/utils/retry.go

```diff
@@ -21,8 +21,36 @@ import (
 	"time"
 
 	"github.com/gravitational/trace"
+	"math/rand"
 )
 
+// JitterFunc is a function which applies random jitter to
+// a duration.  Used to randomize backoff values.  An
+// instance of JitterFunc likely has internal state so
+// concurrent usage must be managed via external mutex.
+type JitterFunc func(time.Duration) time.Duration
+
+// Jitter builds a JitterFunc.  The resulting JitterFunc may
+// have setup costs and therefore should be cached for efficency.
+// An instance of Jitter should be safe for concurrent usage.
+type Jitter func() JitterFunc
+
+// NewJitter returns the default jitter (currently jitters on
+// the range [n/2,n), but this is subject to change).
+func NewJitter() Jitter {
+	return func() JitterFunc {
```

Comment:
Hmm, this seems to be one level deeper than necessary.
Creating a new `rng` for each cloned `Retry` won't make that much difference in behavior.
How about:
```
type Jitter func(time.Duration) time.Duration

func NewJitter() Jitter {
  rng := rand.New(rand.NewSource(time.Now().UnixNano()))
  return func(d time.Duration) time.Duration {
  ...
  }
}
```

Reply:
Can't do it exactly like that since `rand.Source` isn't safe for concurrent usage.  I do like the simpler API though, so might be worth adding a mutex to get there.  Jittered backoff should be low-contention almost by definition :p